### PR TITLE
task4

### PR DESCRIPTION
**Pull request by ihargit**

### CODE REVIEW

**Comment by ihargit** [comment]
+ 4.1 done https://b2yymidh8l.execute-api.eu-west-1.amazonaws.com/prod/products
+ 4.2 done https://b2yymidh8l.execute-api.eu-west-1.amazonaws.com/prod/products
+  4.3 POST done https://b2yymidh8l.execute-api.eu-west-1.amazonaws.com/prod/products
+ 4 FE done https://dm9otfstrcg58.cloudfront.net/

Additional:
+ POST/products, 400 error code, done
+ All lambdas return error 500 status code, done
+ All lambdas do console.log, done

File: product-service/src/db/models/product.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import Joi from 'joi';
+
+const productSchema = Joi.object({
+  id: Joi.string(),
```
Comment (helly-fox) [inline product-service/src/db/models/product.js:4]: Do you need id validation as you do not pass it?

File: product-service/src/post-product.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import { postProductDB } from './db';
+import allowedOrigins from './constants/alloweb-origins';
+
+export const postProduct = async (event) => {
+  console.log('Lambda invocation with event: ', event);
+  const responce = {
```
Comment (helly-fox) [inline product-service/src/post-product.js]: Typo 'response'

File: product-service/src/get-product-by-id.js [hunk 3]
```diff
@@ -16,7 +16,7 @@ export const getProductById = async (event) => {
       }
     }
     const { productId } = event.pathParameters;
-    const product = products.find(({ id }) => id === productId);
+    const product = await getProductByIdDB(productId);
```
Comment (helly-fox) [inline product-service/src/get-product-by-id.js:19]: Seams like error handling is broken https://b2yymidh8l.execute-api.eu-west-1.amazonaws.com/prod/products/1

File: product-service/src/db/models/product.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  title: Joi.string().required(),
+  description: Joi.string(),
+  price: Joi.number().positive().allow(0),
```
Comment (helly-fox) [inline product-service/src/db/models/product.js:7]: I think it is better to update price validation:
{
"id": "04009d9d-5288-4e17-93d8-63daf33a5609",
"title": "Deer",
"description": "Bamby",
"price": null,
"count": 0
},

File: product-service/src/post-product.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    const result = await postProductDB(requestBody);
+    if (result) {
+      responce.body = JSON.stringify(
+        result,
+        null,
+        2
+      );
+    } else {
+      responce.statusCode = 400;
+      responce.body = 'Can not create product';
```
Comment (helly-fox) [inline product-service/src/post-product.js]: I would better return validation error (to understand what was wrong)

**Review by helly-fox — COMMENTED** [review]
Couple of small changes needed

File: product-service/src/db/models/product.js (hunk 1 above)
Comment (ihargit) [reply product-service/src/db/models/product.js:4]: Thank you for your comment) By desing, it's a full picture of 'products' table schema.
And id validation is needed to pass custom id.
For the lesson purpose I used default generation, but prefer to preserve the possibility to pass (and validate) custom id.

File: product-service/src/post-product.js (hunk 2 above)
Comment (ihargit) [reply product-service/src/post-product.js]: Thank you) Fixed

File: product-service/src/post-product.js (hunk 5 above)
Comment (ihargit) [reply product-service/src/post-product.js]: The validation error in returned by JOI and returned with 400 status and message from catch

File: product-service/src/get-product-by-id.js (hunk 3 above)
Comment (ihargit) [reply product-service/src/get-product-by-id.js:19]: This route is beside the task 4, but thank you for the note. It throws 500 error if id is wrong. I believe, I could throw 404 error for this particular case. Should think on how to split them.

File: product-service/src/db/models/product.js (hunk 4 above)
Comment (ihargit) [reply product-service/src/db/models/product.js:7]: Thank you! Fixed.

**Comment by helly-fox** [comment]
Validation massages are not sent 
Screenshot 2020-11-09 at 16 47 23

Still I disagree that validation is valid

Screenshot 2020-11-09 at 16 47 06

+ I do not like the idea of setting defaults (not counting product id) like count = 0 if it is not passed

**Comment by ihargit** [comment]
> Validation massages are not sent
> 
> * I do not like the idea of setting defaults (not counting product id) like count = 0 if it is not passed

Thank you! I added validation message in response. 
image

About defaults: the idea is when I INSERT row in products table, the related row is created in stocks, by trigger. Right after that I can use both tables on their own with no extra preparations - change count number, for example. For that I need numeric default in here.

**Comment by helly-fox** [comment]
> > Validation massages are not sent
> > 
> > * I do not like the idea of setting defaults (not counting product id) like count = 0 if it is not passed
> 
> Thank you! I added validation message in response.
> image
> 
> About defaults: the idea is when I INSERT row in products table, the related row is created in stocks, by trigger. Right after that I can use both tables on their own with no extra preparations - change count number, for example. For that I need numeric default in here.

Is not it a part of body validation to check the count (that it is required and not negative and pass it further)? 
I mean in this case for end user it is better to return created item in case passed properties and product added into DB are different, that s a bit weird behaviour (I tried to add product with count -1 and got success result :) but in DB it is stored as 0)

**Comment by ihargit** [comment]
For now count is validated negative number can't be entered.

**Comment by helly-fox** [comment]
> For now count is validated negative number can't be entered.

Sorry but this is smth stange in this case for me (expected to fail this request)

Screenshot 2020-11-09 at 21 23 21

**Comment by ihargit** [comment]
Sorry, it was for price, added fix, now for count it works too.

**Review by helly-fox — APPROVED** [review]

**Pull request merged**